store.go

```diff
@@ -23,55 +23,101 @@ import (
 	"os"
 	"path/filepath"
 
+	"github.com/piprate/json-gold/ld"
 	"go.etcd.io/bbolt"
 )
 
 // Store is the main interface for storing/finding documents
 type Store interface {
-	// Collection creates or returns a collection.
+	// JSONCollection creates or returns a JSON Collection.
 	// On the db level it's a bucket for the documents and 1 bucket per index.
-	Collection(name string) Collection
-	// Close closes the bbolt DB
+	JSONCollection(name string) Collection
+	// JSONLDCollection creates or returns a JSON-LD Collection.
+	// On the db level it's a bucket for the documents and 1 bucket per index.
+	JSONLDCollection(name string) Collection
+	// Close the bbolt DB
 	Close() error
 }
 
 // Store holds a reference to the bbolt data file and all collections.
 type store struct {
-	db          *bbolt.DB
-	collections map[string]*collection
+	db             *bbolt.DB
+	collections    map[string]Collection
+	documentLoader ld.DocumentLoader
+	// options is used during configuration
+	options bbolt.Options
+}
+
+// StoreOption is the function type for the Store Options
+type StoreOption func(store *store)
+
+// WithoutSync is a store option which signals the underlying bbolt db to skip syncing with disk
+func WithoutSync() StoreOption {
+	return func(store *store) {
+		store.options.NoSync = true
+	}
+}
+
+// WithDocumentLoader overrides the default document loader
+func WithDocumentLoader(documentLoader ld.DocumentLoader) StoreOption {
+	return func(store *store) {
+		store.documentLoader = documentLoader
+	}
+
 }
 
 // NewStore creates a new store.
 // the noSync option disables flushing to disk, ideal for testing and bulk loading
-func NewStore(dbFile string, noSync bool) (Store, error) {
+func NewStore(dbFile string, options ...StoreOption) (Store, error) {
 	err := os.MkdirAll(filepath.Dir(dbFile), os.ModePerm)
 	if err != nil {
 		return nil, err
 	}
 
-	options := *bbolt.DefaultOptions
-	options.NoSync = noSync
+	// store with defaults
+	st := &store{
+		options:        *bbolt.DefaultOptions,
+		collections:    map[string]Collection{},
+		documentLoader: ld.NewDefaultDocumentLoader(nil),
+	}
+
+	// apply options
+	for _, option := range options {
+		option(st)
+	}
 
-	db, err := bbolt.Open(dbFile, boltDBFileMode, &options)
+	st.db, err = bbolt.Open(dbFile, boltDBFileMode, &st.options)
 	if err != nil {
 		return nil, err
 	}
 
-	st := &store{
-		db:          db,
-		collections: map[string]*collection{},
+	return st, nil
+}
+
+func (s *store) JSONCollection(name string) Collection {
+	c, ok := s.collections[name]
+	if !ok {
+		c = &collection{
+			name:           name,
+			db:             s.db,
+			refMake:        defaultReferenceCreator,
+			valueCollector: JSONPathValueCollector,
+		}
+		s.collections[name] = c
 	}
 
-	return st, nil
+	return c
 }
 
-func (s *store) Collection(name string) Collection {
+func (s *store) JSONLDCollection(name string) Collection {
 	c, ok := s.collections[name]
```

Comment:
what if the collection was created earlier as JSON collection, or vice versa?

Reply:
Then you're screwed.